pkg/solana/contract.go

```diff
@@ -58,32 +81,112 @@ func NewTracker(spec OCR2Spec, client *Client, transmitter TransmissionSigner, l
 		client:          client,
 		lggr:            lggr,
 		requestGroup:    &singleflight.Group{},
+		stateLock:       &sync.RWMutex{},
+		ansLock:         &sync.RWMutex{},
+		stateTime:       time.Now(),
+		ansTime:         time.Now(),
+		staleTimeout:    staleTimeout,
 	}
 }
 
-// fetch + decode + store raw state
-func (c *ContractTracker) fetchState(ctx context.Context) error {
-	c.lggr.Debugf("fetch state for account: %s", c.StateID.String())
+// Start polling
+func (c *ContractTracker) Start() error {
+	return c.StartOnce("pollState", func() error {
+		c.done = make(chan struct{})
+		go c.PollState()
+		return nil
+	})
+}
+
+// PollState contains the state and transmissions polling implementation
+func (c *ContractTracker) PollState() {
+	c.lggr.Debugf("Starting state polling for state: %s, transmissions: %s", c.StateID, c.TransmissionsID)
+	ticker := time.NewTicker(c.client.pollingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.done:
+			c.lggr.Debugf("Stopping state polling for state: %s, transmissions: %s", c.StateID, c.TransmissionsID)
+			return
+		case <-ticker.C:
+			// async poll both transmisison + ocr2 states
+			go func() {
+				ctx, cancel := utils.ContextFromChanWithDeadline(c.done, c.client.contextDuration)
+				defer cancel()
+				_, err, shared := c.requestGroup.Do("state", func() (interface{}, error) {
+					return nil, c.fetchState(ctx)
+				})
+				if err != nil {
+					c.lggr.Errorf("error in PollState.fetchState, shared: %t, error %s", shared, err)
+				}
+			}()
+			go func() {
+				ctx, cancel := utils.ContextFromChanWithDeadline(c.done, c.client.contextDuration)
+				defer cancel()
+				// make single flight request
+				_, err, shared := c.requestGroup.Do("transmissions.latest", func() (interface{}, error) {
+					return nil, c.fetchLatestTransmission(ctx)
+				})
+				if err != nil {
+					c.lggr.Errorf("error in PollState.fetchLatestTransmission, shared: %t, error %s", shared, err)
+				}
+			}()
+		}
+	}
+}
 
-	// make single flight request
-	v, err, shared := c.requestGroup.Do("state", func() (interface{}, error) {
-		state, _, err := GetState(ctx, c.client.rpc, c.StateID)
-		return state, err
+// Close stops the polling
+func (c *ContractTracker) Close() error {
+	return c.StopOnce("pollState", func() error {
+		close(c.done)
+		return nil
 	})
+}
 
+// ReadState reads the latest state from memory with mutex and errors if timeout is exceeded
+func (c *ContractTracker) ReadState() (State, solana.PublicKey, error) {
+	c.stateLock.RLock()
+	defer c.stateLock.RUnlock()
+
+	var err error
+	current := time.Now()
+	if current.After(c.stateTime.Add(c.staleTimeout)) {
```

Comment:
Would this read a little easier?
```suggestion
	if time.Since(c.stateTime) > c.staleTimeout {
```